### PR TITLE
py-cupy: Add 11.3.0, 11.4.0, 11.5.0, 11.6.0, 12.0.0, 12.1.0

### PR DESCRIPTION
**Pull request by elliottslaughter**
For the version bounds, I referred mostly to the `setup.py` files in each version, e.g.:

https://github.com/cupy/cupy/blob/v11.6.0/setup.py

https://github.com/cupy/cupy/blob/v12.0.0/setup.py

And also consulted the release notes for CUDA version support: https://github.com/cupy/cupy/releases

### CODE REVIEW

**Review by eugeneswalker — APPROVED** [review]

File: var/spack/repos/builtin/packages/py-cupy/package.py [hunk 1]
```diff
@@ -17,13 +17,23 @@ class PyCupy(PythonPackage, CudaPackage):
...
 
     version("11.2.0", sha256="c33361f117a347a63f6996ea97446d17f1c038f1a1f533e502464235076923e2")
-
-    depends_on("python@3.7:", type=("build", "run"))
+    version("11.3.0", sha256="d057cc2f73ecca06fae8b9c270d9e14116203abfd211a704810cc50a453b4c9e")
+    version("11.4.0", sha256="03d52b2626e02a3a2b46d714c1cd03e702c8fe33915fcca6ed8de5c539964f49")
+    version("11.5.0", sha256="4bc8565bded22cc89b210fd9fb48a5d5316f30701e12bb23852a60314e1f9f6e")
+    version("11.6.0", sha256="53dbb840072bb32d4bfbaa6bfa072365a30c98b1fcd1f43e48969071ad98f1a7")
+    version("12.0.0", sha256="61ddbbef73d50d606bd5087570645f3c91ec9176c2566784c1d486d6a3404545")
+    version("12.1.0", sha256="f6d31989cdb2d96581da12822e28b102f29e254427195c2017eac327869b7320")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-cupy/package.py:25]: It's customary to sort versions from newest to oldest

File: var/spack/repos/builtin/packages/py-cupy/package.py [hunk 2]
```diff
@@ -17,13 +17,23 @@ class PyCupy(PythonPackage, CudaPackage):
...
+    depends_on("python@3.7:", when="@:11", type=("build", "run"))
+    depends_on("python@3.8:", when="@12:", type=("build", "run"))
     depends_on("py-setuptools", type="build")
     depends_on("py-cython@0.29.22:2", type="build")
     depends_on("py-fastrlock@0.5:", type=("build", "run"))
-    depends_on("py-numpy@1.20:1.25", type=("build", "run"))
-    depends_on("cuda")
+    depends_on("py-numpy@1.20:1.25", when="@:11", type=("build", "run"))
+    depends_on("py-numpy@1.20:1.26", when="@12:", type=("build", "run"))
+    depends_on("py-scipy@1.6:1.12", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-cupy/package.py:34]: Scipy is optional, not required. If you want you can add a `+all` variant to control whether or not scipy/cython/optuna get installed, but I would default to False so as to match the behavior of `pip install cupy`.

File: var/spack/repos/builtin/packages/py-cupy/package.py [hunk 3]
```diff
@@ -17,13 +17,23 @@ class PyCupy(PythonPackage, CudaPackage):
...
+    depends_on("cuda@:11.9", when="@:11")
+    depends_on("cuda@:12.1", when="@12:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-cupy/package.py:36]: Can you add a comment like:
```python
# Based on https://github.com/cupy/cupy/releases
```
so we know where to check in the future?

File: var/spack/repos/builtin/packages/py-cupy/package.py (hunk 2 above)
Comment (eugeneswalker) [reply var/spack/repos/builtin/packages/py-cupy/package.py:34]: Addressing your review remarks in follow up PR here:
* https://github.com/spack/spack/pull/38919

**Pull request merged**